dynamic_programming/smith_waterman.py

```diff
@@ -0,0 +1,97 @@
+# https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm
+# Score constants
+"""
+Score constants used in the Smith-Waterman algorithm. Matches are given a positive
+score while mismatches are given a negative score. Gaps are also penalized.
+"""
+MATCH = 1
+MISMATCH = -1
+GAP = -2
+
+
+def score_function(source_char: str, target_char: str) -> int:
+    """
+    Calculate the score for a character pair based on whether they match or mismatch.
+    Returns 1 if the characters match, -1 if they mismatch.
+    >>> score_function('A', 'A')
+    1
+    >>> score_function('A', 'C')
+    -1
+    """
+    if source_char == target_char:
+        return MATCH
+    else:
+        return MISMATCH
+
+
+def smith_waterman(query: str, subject: str) -> list[list[int]]:
+    """
+    Perform the Smith-Waterman local sequence alignment algorithm.
+    Returns a 2D list representing the score matrix. Each value in the matrix
+    corresponds to the score of the best local alignment ending at that point.
+    >>> smith_waterman('ACAC', 'CA')
+    [[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 0, 2], [0, 1, 0]]
```

Comment:
Can you write some more tests testing the following cases:
1. Empty `query`
2. Empty `subject`
3. Lowercase string

Reply:
done